clusterloader/doc/vision.md

```diff
@@ -0,0 +1,369 @@
+We may want to from time to time tweak how do we measure existing SLOs or introduce
+a new one, but it is fine to require changes to the framework to achieve that.
+
+As a result, mechanisms to measure specific SLOs (or gather other types of metrics)
+will be incorporated into Cluster Loader framework. We will expect that developers
+trying to introduce a new SLO (or change how do we measure the existing one) will
+be modifying that part of Cluster Loader codebase. Within the codebase, we will try
+to provide a relatively easy framework to achieve it though.
+
+At the high level, to implement gathering a given portion of data or measure a new
+SLO, you will need to implement a very simple interface, similar to:
+
+```
+type Measurement interface {
+	StartMeasurement(config MeasurementConfig)
+	StopMeasurement()
+}
+
+struct MeasurementConfig {
+	// Client to access a cluster
+	Clientset k8sclient.ClientSet
+	// More needed parameters to be determined.
+	// TBD: Exact method of passing them. One possibility is having
+	// per-type config and a string field here with json-serialized
+	// config that will be understood only by a given measurement
+	// type.
+}
+```
+
+Providing an implementation of such interface and registering it in the correct
+place will allow you to use those as phases in your config.
+As an example, consider gathering resource usage from system components.
+It will be enough to implement the following methods:
+
+```
+struct ResourceGatherer {
+	// Some fields that you need.
+}
+func (r *ResourceGatherer) StartMeasurement(c MeasurementConfig) {
+	// Initialize gatherer and start gathering metrics.
+}
+funct (r *ResourceGatherer) StopMeasurement() {
+	// Stop gatherer.
+	// Validate and/or report results.
+}
+```
+
+and the registering it in some config file, will let you automatically use:
+
+```
+	ResourceGatherer_Start
+	ResourceGatherer_Stop
+```
+
+as “phases” of your test.
+
+At the beginning, we will provide the following measurement types:
+
+```
+	ResourceGatherer_Start/Stop
+	Profiler_Start/Stop
+	MetricsGatherer_Start/Stop
+	APICallLatencyValidator_Start/Stop
+	PodStartupLatencyValidator_Start/Stop
+```
+
+The remaining question is how to incorporate them into config. To avoid
+introducing another fields to “Phase” type (and making it more confusing to users),
+we introduce a concept of a “test step” where a test step can be either a
+number of phases or a number of measurement actions. We define it as following:
+
+```
+struct Step {
+	// Exactly one of these slices should be non-empty.
+	Phases       []Phase
+	Measurements []string
+}
+```
+
+
+### Top level config
+To define a test scenario, user will be providing the following config:
+
+```
+struct Config {
+	// TBD: Way to access cluster.
+
+	// Number of namespace in the test.
+	Namespaces int32
+	// Steps of the test.
+	Steps []Step
+}
+```
+
+With a test being defined by a single json-file, it should be pretty simple to
```

Comment:
Will yaml be compatible, as it is far more human readable?

Reply:
Yes - we should make it work for both json and yaml.

Reply:
Sounds good

Reply:
Done